src/test/java/bowling/FrameTest.java

```diff
@@ -0,0 +1,225 @@
+
+        frame.roll(1);
+        assertThat(frame.isOver(), is(true));
+    }
+
+    @Test
+    public void strikeFrameHasOneRoll() throws BowlingException {
+        Frame frame = new Frame();
+
+        frame.roll(10);
+        assertThat(frame.isOver(), is(true));
+    }
+
+    @Test
+    public void normalFrameIsCorrectlyIdentified() throws BowlingException {
+        Frame frame = new Frame();
+        frame.roll(1);
+        frame.roll(1);
+        assertThat(frame.getFrameType(), is(FrameType.NORMAL));
+    }
+
+    @Test
+    public void spareFrameIsCorrectlyIdentified() throws BowlingException {
+        Frame frame = new Frame();
+        frame.roll(1);
+        frame.roll(9);
+        assertThat(frame.getFrameType(), is(FrameType.SPARE));
+    }
+
+    @Test
+    public void strikeFrameIsCorrectlyIdentified() throws BowlingException {
+        Frame frame = new Frame();
+        frame.roll(10);
+        assertThat(frame.getFrameType(), is(FrameType.STRIKE));
+    }
+
+    @Test
+    public void finalFrameHasThreeRollsForAStrike() throws BowlingException {
+        Frame frame = new Frame(true);
+
+        frame.roll(10);
+        assertThat(frame.isOver(), is(false));
+
+        frame.roll(1);
+        assertThat(frame.isOver(), is(false));
+
+        frame.roll(1);
+        assertThat(frame.isOver(), is(true));
+    }
+
+    @Test
+    public void finalFrameHasThreeRollsForASpare() throws BowlingException {
+        Frame frame = new Frame(true);
+
+        frame.roll(7);
+        assertThat(frame.isOver(), is(false));
+
+        frame.roll(3);
+        assertThat(frame.isOver(), is(false));
+
+        frame.roll(1);
+        assertThat(frame.isOver(), is(true));
+    }
+
+    @Test
+    public void finalFrameAllStrikesScore() throws BowlingException {
+        Frame frame = new Frame(true);
+
+        frame.roll(10);
+        frame.roll(10);
+        frame.roll(10);
+
+        assertThat(frame.score(), is(30));
+    }
+
+    @Test
+    public void finalFrameSpareScore() throws BowlingException {
+        Frame frame = new Frame(true);
+
+        frame.roll(1);
+        frame.roll(9);
+        frame.roll(3);
+
+        assertThat(frame.score(), is(1+9+3));
+    }
+
+    @Test
+    public void finalFrameNormalScore() throws BowlingException {
+        Frame frame = new Frame(true);
+
+        frame.roll(2);
+        frame.roll(2);
+
+        assertThat(frame.score(), is(4));
+    }
+
+    @Test
+    public void bonusIsApplied() throws BowlingException {
+        Frame frame = new Frame();
+        frame.roll(10);
+        frame.addBonus(3);
+        frame.addBonus(4);
+        assertThat(frame.score(), is(10+3+4));
+    }
+
+    @Test
+    public void bonusIsNotAppliedOnFinalFrame() throws BowlingException {
+        Frame frame = new Frame(true);
+        frame.roll(10);
+        frame.addBonus(3);
+        frame.addBonus(4);
+        assertThat(frame.score(), is(10));
+    }
+
+    @Test(expected = FrameOverException.class)
```

Comment:
I realise that this is "the junit way". So 100% valid. But personally I like to make it explicitly clear in the test code that exceptions are being validated against.

```
try {
doSomething()
fail("Expected exception")
} catch (e) {
assert(e, expected)
}
```

This is obviously quite controversial, but its easy to miss `@Test(excepted` or `throws`. Perhaps what you could do is verify the message which would justify catching the exception in such a way. Alternatively I think JUnit has actually introduced better exception assertion handling, might be worth having a look.